src/Migrations.ts

```diff
@@ -0,0 +1,204 @@
+   *
+   * Migrates database schema to the latest version. Filename option is present
+   * in the the MigrationData if `readMigrations` function executes or user can put filename
+   * for each migration in the config.migrations
+   */
+  async migrate (): Promise<void> {
+    const table = this.config.table
+    const migrations = this.config.migrations ?? (await this.readMigrations())
+
+    // Create a database table for migrations meta data if it doesn't exist
+    await this.db.run(`CREATE TABLE IF NOT EXISTS "${table}" (
+      id   INTEGER PRIMARY KEY,
+      name TEXT    NOT NULL,
+      up   TEXT    NOT NULL,
+      down TEXT    NOT NULL
+    )`)
+
+    // Get the list of already applied migrations
+    let dbMigrations = await this.db.all(
+      `SELECT id, name, up, down FROM "${table}" ORDER BY id ASC`
+    )
+
+    await this.downMigrations(dbMigrations, migrations)
+    await this.upNewMigrations(dbMigrations, migrations)
+  }
+
+  /**
+   * Apply pending migrations
+   *
+   * @param dbMigrations exists migrations in the `table`
+   * @param migrations IMigrate.MigrationData
+   */
+  private async upNewMigrations (
+    dbMigrations: any[],
+    migrations: readonly MigrationData[]
+  ): Promise<void> {
+    const lastMigrationId = dbMigrations.length
+      ? dbMigrations[dbMigrations.length - 1].id
+      : 0
+    for (const migration of migrations) {
+      if (migration.id > lastMigrationId) {
+        await this.db.run('BEGIN')
+        try {
+          await this.db.exec(migration.up)
+          await this.db.run(
+            `INSERT INTO "${this.config.table}" (id, name, up, down) VALUES (?, ?, ?, ?)`,
+            migration.id,
+            migration.name,
+            migration.up,
+            migration.down
+          )
+          await this.db.run('COMMIT')
+        } catch (err) {
+          await this.db.run('ROLLBACK')
+          throw err
+        }
+      }
+    }
+  }
+
+  /**
+   * Undo migrations that exist only in the database but not in files,
+   * also undo the last migration if the `force` option is enabled.
+   *
+   * @param dbMigrations exists migrations in the `table`
+   * @param migrations IMigrate.MigrationData
+   */
+  private async downMigrations (
+    dbMigrations: any[],
+    migrations: readonly MigrationData[]
+  ): Promise<void> {
+    const lastMigration = migrations[migrations.length - 1]
+    for (const migration of dbMigrations
+      .slice()
+      .sort((a, b) => Math.sign(b.id - a.id))) {
+      if (
+        !migrations.some(x => x.id === migration.id) ||
+        (this.config.force && migration.id === lastMigration.id)
+      ) {
+        await this.db.run('BEGIN')
+        try {
+          await this.db.exec(migration.down)
+          await this.db.run(
+            `DELETE FROM "${this.config.table}" WHERE id = ?`,
+            migration.id
+          )
+          await this.db.run('COMMIT')
+          dbMigrations = dbMigrations.filter(x => x.id !== migration.id)
```

Comment:
Why do we need this one?
As I understand it'll remove downed migrations and pass migrations that exists to the upMigrations function

But I tried to run this code and got:
<details>
<summary> Logs </summary>
 npx ts-node-dev  test.ts                                                                                                                                                              14:36:46 ∞
[INFO] 14:36:55 ts-node-dev ver. 1.1.8 (using ts-node ver. 9.1.1, typescript ver. 4.6.3)
[
  {
    id: 1,
    name: 'initial',
    up: 'CREATE TABLE Category (\n' +
      '  id   INTEGER PRIMARY KEY,\n' +
      '  name TEXT    NOT NULL\n' +
      ');\n' +
      '\n' +
      'CREATE TABLE Post (\n' +
      '  id          INTEGER PRIMARY KEY,\n' +
      '  categoryId  INTEGER NOT NULL,\n' +
      '  title       TEXT    NOT NULL,\n' +
      '  isPublished NUMERIC NOT NULL DEFAULT 0,\n' +
      '  CONSTRAINT Post_fk_categoryId FOREIGN KEY (categoryId)\n' +
      '    REFERENCES Category (id) ON UPDATE CASCADE ON DELETE CASCADE,\n' +
      '  CONSTRAINT Post_ck_isPublished CHECK (isPublished IN (0, 1))\n' +
      ');\n' +
      '\n' +
      'CREATE INDEX Post_ix_categoryId ON Post (categoryId);\n' +
      '\n' +
      "INSERT INTO Category (id, name) VALUES (1, 'Test');",
    down: 'DROP INDEX Post_ix_categoryId;\nDROP TABLE Post;\nDROP TABLE Category;'
  },
  {
    id: 2,
    name: 'some-feature',
    up: 'CREATE TABLE Test (\n' +
      '  id   INTEGER PRIMARY KEY,\n' +
      '  name TEXT    NOT NULL\n' +
      ');',
    down: 'DROP TABLE Test;'
  },
  {
    id: 3,
    name: 'test-cert',
    up: 'CREATE TABLE whatever ( certificate TEXT );\n' +
      'INSERT INTO whatever ( certificate ) VALUES (\n' +
      "  '-----BEGIN CERTIFICATE-----\n" +
      'some contents\n' +
      "-----END CERTIFICATE-----');",
    down: 'DROP TABLE whatever;'
  },
  {
    id: 4,
    name: 'no-down',
    up: 'CREATE TABLE IF NOT EXISTS downless ( value TEXT );\n' +
      "INSERT INTO downless ( value ) VALUES ('down migration is optional');",
    down: ''
  }
]
[
  {
    id: 1,
    name: 'initial',
    up: 'CREATE TABLE Category (\n' +
      '  id   INTEGER PRIMARY KEY,\n' +
      '  name TEXT    NOT NULL\n' +
      ');\n' +
      '\n' +
      'CREATE TABLE Post (\n' +
      '  id          INTEGER PRIMARY KEY,\n' +
      '  categoryId  INTEGER NOT NULL,\n' +
      '  title       TEXT    NOT NULL,\n' +
      '  isPublished NUMERIC NOT NULL DEFAULT 0,\n' +
      '  CONSTRAINT Post_fk_categoryId FOREIGN KEY (categoryId)\n' +
      '    REFERENCES Category (id) ON UPDATE CASCADE ON DELETE CASCADE,\n' +
      '  CONSTRAINT Post_ck_isPublished CHECK (isPublished IN (0, 1))\n' +
      ');\n' +
      '\n' +
      'CREATE INDEX Post_ix_categoryId ON Post (categoryId);\n' +
      '\n' +
      "INSERT INTO Category (id, name) VALUES (1, 'Test');",
    down: 'DROP INDEX Post_ix_categoryId;\nDROP TABLE Post;\nDROP TABLE Category;'
  },
  {
    id: 2,
    name: 'some-feature',
    up: 'CREATE TABLE Test (\n' +
      '  id   INTEGER PRIMARY KEY,\n' +
      '  name TEXT    NOT NULL\n' +
      ');',
    down: 'DROP TABLE Test;'
  },
  {
    id: 3,
    name: 'test-cert',
    up: 'CREATE TABLE whatever ( certificate TEXT );\n' +
      'INSERT INTO whatever ( certificate ) VALUES (\n' +
      "  '-----BEGIN CERTIFICATE-----\n" +
      'some contents\n' +
      "-----END CERTIFICATE-----');",
    down: 'DROP TABLE whatever;'
  }
]
[
  {
    id: 1,
    name: 'initial',
    up: 'CREATE TABLE Category (\n' +
      '  id   INTEGER PRIMARY KEY,\n' +
      '  name TEXT    NOT NULL\n' +
      ');\n' +
      '\n' +
      'CREATE TABLE Post (\n' +
      '  id          INTEGER PRIMARY KEY,\n' +
      '  categoryId  INTEGER NOT NULL,\n' +
      '  title       TEXT    NOT NULL,\n' +
      '  isPublished NUMERIC NOT NULL DEFAULT 0,\n' +
      '  CONSTRAINT Post_fk_categoryId FOREIGN KEY (categoryId)\n' +
      '    REFERENCES Category (id) ON UPDATE CASCADE ON DELETE CASCADE,\n' +
      '  CONSTRAINT Post_ck_isPublished CHECK (isPublished IN (0, 1))\n' +
      ');\n' +
      '\n' +
      'CREATE INDEX Post_ix_categoryId ON Post (categoryId);\n' +
      '\n' +
      "INSERT INTO Category (id, name) VALUES (1, 'Test');",
    down: 'DROP INDEX Post_ix_categoryId;\nDROP TABLE Post;\nDROP TABLE Category;'
  },
  {
    id: 2,
    name: 'some-feature',
    up: 'CREATE TABLE Test (\n' +
      '  id   INTEGER PRIMARY KEY,\n' +
      '  name TEXT    NOT NULL\n' +
      ');',
    down: 'DROP TABLE Test;'
  },
  {
    id: 3,
    name: 'test-cert',
    up: 'CREATE TABLE whatever ( certificate TEXT );\n' +
      'INSERT INTO whatever ( certificate ) VALUES (\n' +
      "  '-----BEGIN CERTIFICATE-----\n" +
      'some contents\n' +
      "-----END CERTIFICATE-----');",
    down: 'DROP TABLE whatever;'
  },
  {
    id: 4,
    name: 'no-down',
    up: 'CREATE TABLE IF NOT EXISTS downless ( value TEXT );\n' +
      "INSERT INTO downless ( value ) VALUES ('down migration is optional');",
    down: ''
  }
]
</details>

<details>
<summary> test.ts </summary>
import * as sqlite3 from 'sqlite3'

import { Database } from "./src";
import { Migrations } from "./src/Migrations";


const boot = async () => {
  const db = new Database({
      filename: '.db',
      driver: sqlite3.Database,
  })
  
  await db.open()
  
  const instance = new Migrations(db, {force: true})
  
  await instance.migrate();
}

boot()
</details>

Reply:
tests may have to be altered / rewritten / added as the new methods are created